### PR TITLE
tabletserver: SBR deprecation

### LINKED ISSUE #5920: RFC: RBR vttablet -> Breaking changes
## Summary
The deprecation of SBR will result in a fundamental change about how vttablet works. The most significant one is that vttablet does not need to compute which rows will be affected by a DML. Dropping this requirement allows us to rethink the implementation of its various features. Some of these changes will not be fully backward compatible. Here are the details:

## Pass-through DMLs
Most DMLs can be just passed through.

There is, however, one valuable feature that we want to preserve: the ability to limit the number of rows a DML affects. In SBR mode, this was achieved as a side-effect of the fact that we had to run a “subquery” to identify all affected rows, which allowed us to count them and return an error if they exceeded the limit. We lose this ability in pass-through mode.

Instead of continuing to issue the subquery, the new proposal is to add a LIMIT clause to the DML itself. Then, if “RowsAffected” was greater than the limit, we return an error. The one downside of this implementation is that such a failure will leave the DML partially executed. This means that vttablet will have to force a rollback of the transaction. In most use cases, the application is expected to rollback the transaction on such an error. Therefore, this sounds like a reasonable trade-off for the level of simplicity achieved. There is also the added benefit of avoiding the extra roundtrip for the subquery.

This change has a subtle interaction with the “found rows” flag, which reports affected rows differently. This just means that users of this flag may have to set their limits differently.

The explicit pass-through mode flag `queryserver-config-passthrough-dmls` will continue to behave as is. In the new context, its meaning will change to: “do not limit affected rows”.

We believe that inserts of a large number of rows are usually intentional. Therefore, limits will only be applied to updates and deletes.

## Autocommits
With most DMLs being pass-through, we can now resurrect autocommit behavior in vttablet. This means that a transactionless DML can be treated as autocommit. In the case where a limit has to be enforced, vttablet can open a transaction, and then rollback if the limit is exceeded.

## Sequences
Sequences will continue their existing behavior: normal statements will continue to be sent to mysql. If a `select next…` is received, the sequence specific functionality will be triggered.

## Messages
Messages are undergoing a significant overhaul as seen in #5913. Tracking of row changes has been changed to use VStreamer, which allows for any DML to be executed on those tables. Inserts remain an exception because columns like `time_created` and `time_scheduled` have to be populated by vttablet in order to maintain backward compatibility.

## Plan IDs
Most plan ids were specific to SBR. After the refactor, the total number of plans will be greatly reduced, and we’ll likely generate a new set of ids.

## Schema
These changes greatly reduce our dependence on the schema. We only need to know the following info from a table:
* Field info
* PK columns
* Table comments

VTTablet also gathered table statistics from `information schema`. However, we don’t think anyone is using them. So, it may be better to stop reporting them. If anyone still needs them, please speak up now.

### PR DESCRIPTION
**Pull request by sougou**
Closes #5920 

The planbuilder and query_executor were essentially rewritten. I had to make tweaks in a few other places to accommodate changes in behavior.

### CODE REVIEW

File: go/vt/vttablet/tabletserver/planbuilder/plan.go [hunk 1]
```diff
@@ -43,60 +40,47 @@ var (
 // PlanType indicates a query plan type.
 type PlanType int
 
+// The following are PlanType values.
```
Comment (systay) [inline go/vt/vttablet/tabletserver/planbuilder/plan.go]: Most of these are self-evident, but for a few of them, I would not mind comments with explanation

File: go/vt/vttablet/tabletserver/query_executor_test.go [hunk 2]
```diff
@@ -47,982 +48,366 @@ import (
...
+			tsv := newTestTabletServer(ctx, smallResultSize, db)
+			defer tsv.StopService()
+
+			tsv.SetPassthroughDMLs(tcase.passThrough)
+
+			// Test outside a transaction.
+			qre := newTestQueryExecutor(ctx, tsv, tcase.input, 0)
+			_, err := qre.Execute()
+			wantErr := "caller id: d: row count exceeded 2 (errno 10001) (sqlstate HY000)"
+			if err == nil || err.Error() != wantErr {
```
Comment (systay) [inline go/vt/vttablet/tabletserver/query_executor_test.go]: nit: You could use
```
require.Error(t, wantErr)
```

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 3]
```diff
@@ -120,84 +114,102 @@ func (qre *QueryExecutor) Execute() (reply *sqltypes.Result, err error) {
...
-			// handled above:
-			// planbuilder.PlanNextval
-			// planbuilder.PlanDDL
-
-			// not valid for Execute:
-			// planbuilder.PlanSelectStream
-			// planbuilder.PlanMessageStream:
-			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "%s unexpected plan type", qre.plan.PlanID.String())
+		if err := qre.verifyRowCount(int64(len(qr.Rows)), maxrows); err != nil {
+			return nil, err
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go]: Test exists for DML Limit.  Not able to find a test for select limit.

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 4]
```diff
@@ -97,18 +97,12 @@ func (qre *QueryExecutor) Execute() (reply *sqltypes.Result, err error) {
...
 			return &sqltypes.Result{
-				Fields:       qre.plan.Fields,
-				RowsAffected: 0,
-				InsertID:     0,
-				Rows:         nil,
-				Extras:       nil,
+				Fields: qre.plan.Fields,
 			}, nil
 		}
 	}
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go]: case planbuilder.PlanSelect, planbuilder.PlanSelectImpossible also exists below. Why "planbuilder.PlanSelectImpossible" is present in two switch statement?

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 5]
```diff
@@ -120,84 +114,102 @@ func (qre *QueryExecutor) Execute() (reply *sqltypes.Result, err error) {
...
+	if err != nil {
+		// dbConn is nil, it means the transaction was aborted.
+		// If so, we should not relog the rollback.
+		// TODO(sougou): these txPool functions should take the logstats
+		// and log any statements they issue. This needs to be done as
+		// a separate refactor because it impacts lot of code.
+		if conn.dbConn != nil {
+			defer qre.logStats.AddRewrittenSQL("rollback", time.Now())
+			qre.tsv.te.txPool.LocalConclude(qre.ctx, conn)
+		}
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go]: I think test is missing for this condition.

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 6]
```diff
@@ -761,12 +542,12 @@ func (qre *QueryExecutor) getStreamConn() (*connpool.DBConn, error) {
...
 func (qre *QueryExecutor) qFetch(logStats *tabletenv.LogStats, parsedQuery *sqlparser.ParsedQuery, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
-	sql, sqlWithoutComments, err := qre.generateFinalSQL(parsedQuery, bindVars, nil, "")
+	sql, sqlWithoutComments, err := qre.generateFinalSQL(parsedQuery, bindVars)
 	if err != nil {
 		return nil, err
 	}
 	// Check tablet type.
-	if qre.tsv.qe.enableConsolidator || (qre.tsv.qe.enableConsolidatorReplicas && qre.tabletType != topodata.TabletType_MASTER) {
+	if qre.tsv.qe.enableConsolidator || (qre.tsv.qe.enableConsolidatorReplicas && qre.tabletType != topodatapb.TabletType_MASTER) {
 		q, original := qre.tsv.qe.consolidator.Create(string(sqlWithoutComments))
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go]: consolidator is default to true and hence test does not go in else part. Can a test be added?

File: go/vt/vttablet/tabletserver/planbuilder/plan.go [hunk 7]
```diff
@@ -274,7 +170,7 @@ func Build(statement sqlparser.Statement, tables map[string]*schema.Table) (*Pla
 	case *sqlparser.Set:
 		plan, err = analyzeSet(stmt), nil
 	case *sqlparser.DDL:
-		plan, err = analyzeDDL(stmt, tables), nil
+		plan = &Plan{PlanID: PlanDDL}
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/planbuilder/plan.go]: This does not contain the DDL Query. Could you explain the reason for not adding the query?

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 3 above)
Comment (sougou) [reply go/vt/vttablet/tabletserver/query_executor.go]: Good catch. I must have accidentally deleted an existing test.

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 6 above)
Comment (sougou) [reply go/vt/vttablet/tabletserver/query_executor.go]: This test requires creating a delay in query execution, which the test framework can't emulate. So, it's covered in the `go/vt/vttablet/endtoend/misc_test.go`.

File: go/vt/vttablet/tabletserver/query_executor_test.go (hunk 2 above)
Comment (sougou) [reply go/vt/vttablet/tabletserver/query_executor_test.go]: For some reason, assert didn't work. In any case, I had to change the test to use "Contains".
But `assert.Contains` can't check if an error contains a string, which is unfortunate.

File: go/vt/vttablet/tabletserver/tabletenv/config.go [hunk 8]
```diff
@@ -64,7 +66,7 @@ func init() {
 	flag.IntVar(&Config.WarnResultSize, "queryserver-config-warn-result-size", DefaultQsConfig.WarnResultSize, "query server result size warning threshold, warn if number of rows returned from vttablet for non-streaming queries exceeds this")
 	flag.IntVar(&Config.MaxDMLRows, "queryserver-config-max-dml-rows", DefaultQsConfig.MaxDMLRows, "query server max dml rows per statement, maximum number of rows allowed to return at a time for an update or delete with either 1) an equality where clauses on primary keys, or 2) a subselect statement. For update and delete statements in above two categories, vttablet will split the original query into multiple small queries based on this configuration value. ")
 	flag.BoolVar(&Config.PassthroughDMLs, "queryserver-config-passthrough-dmls", DefaultQsConfig.PassthroughDMLs, "query server pass through all dml statements without rewriting")
-	flag.BoolVar(&Config.AllowUnsafeDMLs, "queryserver-config-allowunsafe-dmls", DefaultQsConfig.AllowUnsafeDMLs, "query server allow unsafe dml statements")
+	flag.BoolVar(&deprecateAllowUnsafeDMLs, "queryserver-config-allowunsafe-dmls", false, "deprecated")
```
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vttablet/tabletserver/tabletenv/config.go]: Unsafe DMLs are those that were considered unsafe for SBR, like "insert on duplicate key", etc.

File: go/vt/vttablet/tabletserver/tabletserver.go [hunk 9]
```diff
@@ -881,7 +835,7 @@ func (tsv *TabletServer) CreateTransaction(ctx context.Context, target *querypb.
 	return tsv.execRequest(
 		ctx, tsv.QueryTimeout.Get(),
 		"CreateTransaction", "create_transaction", nil,
-		target, nil, true /* isBegin */, true, /* allowOnShutdown */
+		target, nil, true, /* allowOnShutdown */
```
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vttablet/tabletserver/tabletserver.go]: It either doesn't matter, or it's extremely important that it be true. Changing to false is likely risky without knowing why.

File: proto/query.proto [hunk 10]
```diff
@@ -367,11 +349,13 @@ message ResultExtras {
 // len(QueryResult[0].fields) is always equal to len(row) (for each
 // row in rows for each QueryResult in QueryResult[1:]).
 message QueryResult {
+  // This used to be ResultExtras.
+  reserved 5;
+
   repeated Field fields = 1;
   uint64 rows_affected = 2;
   uint64 insert_id = 3;
   repeated Row rows = 4;
```
[the remark this answers is not in the document]
Comment (sougou) [reply proto/query.proto]: Better to follow the rules. Then we don't have to worry about corner cases.

**Pull request merged**